### PR TITLE
Fix MachinePool scale-down with unready instances

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

MachinePool scale-down currently counts only ready instances, so an extra VM can stick around forever if it never becomes ready. This makes explicit scale-down count all active instances and prefer removing unready ones first.

**Which issue(s) this PR fixes**:

None.

**Special notes for your reviewer**:

Includes regression coverage for the reported scale-up/scale-down failure.

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] cherry-pick candidate

**Release note**:

```release-note
Fix MachinePool scale-down when excess VMSS instances are not ready.
```

### CODE REVIEW

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go [hunk 1]
```diff
@@ -176,16 +178,20 @@ func (rollingUpdateStrategy rollingUpdateStrategy) SelectMachinesToDelete(ctx co
...
+
+		// Prefer deleting unready machines when explicitly scaling down. Counting only ready machines here can
+		// leave the pool permanently over-provisioned if an extra VM joins the cluster but never becomes ready.
+		for _, v := range unreadyMachines {
+			if len(toDelete) >= overProvisionCount {
+				return toDelete, nil
+			}
+
+			toDelete = append(toDelete, v)
+		}
```
Comment (willie-yao) [inline azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: This also deletes an unready maxSurge replacement during a rolling update. With desired replicas 2, two ready old-model machines, and one unready latest-model machine, overProvisionCount becomes 1 and this loop selects the replacement. The rollout can repeatedly delete its new VM and never progress. Please preserve the ready-count gate while old-model machines remain, and add a regression test for this state.

Comment (mboersma) [reply azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: Good catch. Updated this so that while old-model machines remain, only ready replacements count as excess capacity. Once everything is on the latest model, active unready instances count for explicit scale-down. I also added the exact 2 ready old-model + 1 unready latest-model maxSurge regression case. The focused package tests pass.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go [hunk 2]
```diff
@@ -137,9 +137,18 @@ func (rollingUpdateStrategy rollingUpdateStrategy) SelectMachinesToDelete(ctx co
...
 		deletingMachines           = order(getDeletingMachines(machinesByProviderID))
 		readyMachines              = order(getReadyMachines(machinesByProviderID))
+		unreadyMachines            = order(getUnreadyMachines(machinesByProviderID))
 		machinesWithoutLatestModel = order(getMachinesWithoutLatestModel(machinesByProviderID))
-		overProvisionCount         = len(readyMachines) - int(desiredReplicaCount)
-		disruptionBudget           = func() int {
+		overProvisionCount         = func() int {
+			// During a rollout, only ready replacements count as excess capacity. Counting an unready surge
+			// replacement here can cause it to be repeatedly deleted before the rollout can progress.
+			if len(machinesWithoutLatestModel) > 0 {
```
Comment (willie-yao) [inline azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: This check assumes that any old-model machine means a rollout is in progress, but that is not always true. When `SkipMachinePoolModelReconciliation` is enabled, old-model machines are allowed to remain until the pool is explicitly scaled. In that case, a pool with two ready machines and one extra unready machine still calculates zero excess machines and fails to scale down. The same problem occurs when the extra unready machine is itself on the old model. Please protect an unready machine only when it is a new-model replacement for a ready old-model machine during a rollout, and add tests for these cases.

Comment (mboersma) [reply azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: Good catch. I narrowed the rollout protection to pair ready old-model machines with unready latest-model replacements, and disabled it when SkipMachinePoolModelReconciliation is enabled. I also added coverage for that gate and for deleting an excess unready old-model machine.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go [hunk 3]
```diff
@@ -302,6 +367,23 @@ func getReadyMachines(machinesByProviderID map[string]infrav1exp.AzureMachinePoo
 	return readyMachines
 }
 
+func getUnreadyMachines(machinesByProviderID map[string]infrav1exp.AzureMachinePoolMachine) []infrav1exp.AzureMachinePoolMachine {
+	var unreadyMachines []infrav1exp.AzureMachinePoolMachine
+	for _, v := range machinesByProviderID {
+		if v.Status.Ready || !v.DeletionTimestamp.IsZero() {
+			continue
+		}
+		if v.Status.ProvisioningState != nil &&
```
Comment (willie-yao) [inline azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: This includes instances whose status has not been observed yet. A newly created AzureMachinePoolMachine starts with `ProvisioningState == nil` and `LatestModelApplied == false`, so during a rollout it is treated as an unready old-model machine and selected for deletion before its model status is populated. That can reintroduce the replacement-deletion loop from the earlier review. Please exclude instances with unknown provisioning state and add a test with two ready old-model machines plus one newly created machine with empty status.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go [hunk 4]
```diff
@@ -176,16 +204,41 @@ func (rollingUpdateStrategy rollingUpdateStrategy) SelectMachinesToDelete(ctx co
...
+		}
+		unreadyLatestModelDeleteCount := 0
+		for _, v := range unreadyMachines {
+			if v.Status.LatestModelApplied {
+				unreadyLatestModelDeleteCount++
+			}
+		}
+		unreadyLatestModelDeleteCount -= protectedUnreadyCount
+		for _, v := range unreadyMachines {
+			if len(toDelete) >= overProvisionCount || unreadyLatestModelDeleteCount <= 0 {
```
Comment (willie-yao) [inline azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: This should stop scanning unready latest-model machines, not return from the function. If all of them are protected but there is separate ready excess, this returns fewer candidates than `overProvisionCount`. For example, with desired replicas 2, two ready old-model machines, one ready latest-model machine, and one protected unready latest-model replacement, `overProvisionCount` is 1, but this returns an empty slice instead of deleting a ready old-model machine. Please continue through the ready candidates and add this regression case.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go [hunk 5]
```diff
@@ -176,16 +204,41 @@ func (rollingUpdateStrategy rollingUpdateStrategy) SelectMachinesToDelete(ctx co
...
-
 	// we have too many machines, let's choose the oldest to remove
 	if overProvisionCount > 0 {
 		var toDelete []infrav1exp.AzureMachinePoolMachine
-		log.Info("over-provisioned", "desiredReplicaCount", desiredReplicaCount, "overProvisionCount", overProvisionCount, "machinesWithoutLatestModel", getProviderIDs(machinesWithoutLatestModel), "skipModelReconciliation", skipModelReconciliation)
+		log.Info("over-provisioned", "desiredReplicaCount", desiredReplicaCount, "overProvisionCount", overProvisionCount, "unreadyMachines", getProviderIDs(unreadyMachines), "machinesWithoutLatestModel", getProviderIDs(machinesWithoutLatestModel), "skipModelReconciliation", skipModelReconciliation)
+
+		// Prefer deleting unready machines when explicitly scaling down. Counting only ready machines here can
+		// leave the pool permanently over-provisioned if an extra VM joins the cluster but never becomes ready.
+		for _, v := range unreadyMachines {
```
Comment (willie-yao) [inline azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: This still prioritizes stale-model machines when `SkipMachinePoolModelReconciliation` is enabled. With that gate enabled, model age should not override the delete policy. For example, `Oldest` currently selects a newer stale unready machine before an older latest-model unready machine. Please preserve the configured ordering among unready machines when model reconciliation is skipped and add a mixed-model test.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go (hunk 4 above)
Comment (mboersma) [reply azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: Fixed—this now continues on to ready candidates, with the example case covered.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go (hunk 5 above)
Comment (mboersma) [reply azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: Yep. With the gate enabled, mixed-model unready machines now preserve the configured delete ordering.

File: azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go (hunk 3 above)
Comment (mboersma) [reply azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go]: Good point. Unknown provisioning states are now excluded, with the exact rollout case covered.

**Review by willie-yao — APPROVED** [review]
/lgtm
/approve

Thanks for the fix! 🚀

**Comment by nilo19** [comment]
Will this one be cherry picked?
